File: My_Notes/Programming_Note/Samples/async_samples/ts/engineer.ts
import "./job"

export interface IEngineer {
    do(job: Job): void;
}

export class DBEngineer implements IEngineer {
    async do(job:Job){
        console.log("db task start")
        await new Promise(f => setTimeout(f, 1000 * job.db));
        console.log("db task complete")
    }
}

export class APIEngineer implements IEngineer {
    async do(job:Job){
        console.log("api task start")
        await new Promise(f => setTimeout(f, 1000 * job.api));
        console.log("api task complete")
    }
}

export class UIEngineer implements IEngineer {
    async do(job:Job){
        console.log("ui task start")
        await new Promise(f => setTimeout(f, 1000 * job.ui));
        console.log("ui task complete")
    }
}

File: My_Notes/Programming/Samples/async_samples/dart/engineer.dart
import './job.dart';
import 'dart:io';

abstract class IEngineer {
  void doTask(Job job);
  void doTaskAsync(Job job);
}

class DBEngineer extends IEngineer {
  void doTask(Job job) {
    print("db task start");
    sleep(Duration(seconds: job.db));
    print("db task complete");
  }

  Future<void> doTaskAsync(Job job) async {
    print("db task start");
    await Future.delayed(Duration(seconds: job.db));
    print("db task complete");
  }
}

class UIEngineer extends IEngineer {
  void doTask(Job job) {
    print("ui task start");
    sleep(Duration(seconds: job.ui));
    print("ui task complete");
  }

  Future<void> doTaskAsync(Job job) async {
    print("ui task start");
    await Future.delayed(Duration(seconds: job.ui));
    print("ui task complete");
  }
}

class APIEngineer extends IEngineer {
  void doTask(Job job) {
    print("api task start");
    sleep(Duration(seconds: job.api));
    print("api task complete");
  }

  Future<void> doTaskAsync(Job job) async {
    print("api task start");
    await Future.delayed(Duration(seconds: job.api));
    print("api task complete");
  }
}
